test_conformance/compiler/test_feature_macro.cpp

```diff
@@ -0,0 +1,750 @@
+        return TEST_FAIL;
+    }
+
+    if (full_profile)
+    {
+        api_status = CL_TRUE;
+    }
+    else
+    {
+        if (is_extension_available(deviceID, "cles_khr_int64"))
+        {
+            api_status = CL_TRUE;
+        }
+        else
+        {
+            cl_bool double_supported = CL_FALSE;
+            error = check_api_feature_info_capabilities<cl_device_fp_config>(
+                deviceID, context, double_supported, CL_DEVICE_DOUBLE_FP_CONFIG,
+                CL_FP_FMA | CL_FP_ROUND_TO_NEAREST | CL_FP_INF_NAN
+                    | CL_FP_DENORM);
+            test_error(error, "checking CL_DEVICE_DOUBLE_FP_CONFIG failed");
+            if (double_supported == CL_FALSE)
+            {
+                api_status = CL_FALSE;
+            }
+            else
+            {
+                log_error("FP double type is supported and cles_khr_int64 "
+                          "extension not supported\n");
+                return TEST_FAIL;
+            }
+        }
+    }
+
+    error = check_compiler_feature_info(deviceID, context, test_macro_name,
+                                        compiler_status);
+    if (error != CL_SUCCESS)
+    {
+        return error;
+    }
+
+    return feature_macro_verify_results(test_macro_name, api_status,
+                                        compiler_status, supported);
+}
+
+int test_consistency_c_features_list(cl_device_id deviceID,
+                                     std::vector<std::string> vec_to_cmp)
+{
+    log_info("\nComparison list of features: CL_DEVICE_OPENCL_C_FEATURES vs "
+             "API/compiler queries.\n");
+    cl_int error;
+    size_t config_size;
+    std::vector<cl_name_version> vec_device_feature;
+    std::vector<std::string> vec_device_feature_names;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_OPENCL_C_FEATURES, 0, NULL,
+                            &config_size);
+
+    test_error(
+        error,
+        "clGetDeviceInfo asking for CL_DEVICE_OPENCL_C_FEATURES failed.\n");
+    if (config_size == 0)
+    {
+        log_info("Empty list of CL_DEVICE_OPENCL_C_FEATURES returned by "
+                 "clGetDeviceInfo on this device.\n");
+    }
+    else
+    {
+        int vec_elements = config_size / sizeof(cl_name_version);
+        vec_device_feature.resize(vec_elements);
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_OPENCL_C_FEATURES,
+                                config_size, vec_device_feature.data(), 0);
+        test_error(
+            error,
+            "clGetDeviceInfo asking for CL_DEVICE_OPENCL_C_FEATURES failed.\n");
+    }
+    for (auto each_f : vec_device_feature)
+    {
+        vec_device_feature_names.push_back(each_f.name);
+    }
+    sort(vec_to_cmp.begin(), vec_to_cmp.end());
+    sort(vec_device_feature_names.begin(), vec_device_feature_names.end());
+
+    cl_bool result =
+        std::equal(vec_device_feature_names.begin(),
+                   vec_device_feature_names.end(), vec_to_cmp.begin());
```

Comment:
Because this checks the entire range of `vec_device_feature_names` and compares each element of `vec_to_cmp` against it, if `vec_to_cmp` is the longer list then then will pass even though the lists clearly aren't equal.

Swapping the names doesn't fix that because what if `vec_device_feature_names` is longer.

An example of genuine output that clearly contains a problem, but still PASSED:

```
Comparison list of features: CL_DEVICE_OPENCL_C_FEATURES vs API/compiler queries.
Empty list of CL_DEVICE_OPENCL_C_FEATURES returned by clGetDeviceInfo on this device.
Comparison list of features - passed
Supported features based on CL_DEVICE_OPENCL_C_FEATURES API query:

Supported features based on queries to API/compiler :
__opencl_c_fp64
__opencl_c_int64
features_macro passed
PASSED sub-test.
PASSED test.
```

If I switch this to

```suggestion
    cl_bool result = vec_device_feature_names == vec_to_cmp;
```

the test then fails as you would expect.